test/distributed/fsdp/test_fsdp_optim_state.py

```diff
@@ -1575,24 +1578,39 @@ def __init__(self):
             def forward(self, x):
                 return self.net4(self.net3(self.net2(self.net1(x))))
 
-        models = []
-        optims = []
-        state_dicts = []
-        models.append(FSDP(TestDummyModel().cuda(), use_orig_params=True))
-        optims.append(torch.optim.Adam(models[-1].parameters(), lr=1e-2))
-        models.append(FSDP(TestDummyModel().cuda(), use_orig_params=True))
-        optims.append(
+        class FakeMPModel(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                torch.manual_seed(0)
+                self.dense = FSDP(DenseModel().cuda(), use_orig_params=True)
+                if dist.get_rank() == 0:
+                    self.sparse0 = nn.Sequential(nn.Linear(8, 8), nn.ReLU())
+                else:
+                    self.sparse1 = nn.Sequential(nn.Linear(8, 8), nn.ReLU())
+
+            def forward(self, x):
+                if dist.get_rank() == 0:
+                    sparse = self.sparse0(x)
+                else:
+                    sparse = self.sparse1(x)
+                dist.all_reduce(sparse)
+                return self.dense(sparse)
+
+        models = [FakeMPModel().cuda(), FakeMPModel().cuda()]
```

Comment:
Do we need to test wrapping the entire FakeModel with FSDP and telling FSDP to ignore the sparse0 / sparse1? Wondering which approach most accurately simulates the FSDP / TRec integration case.

Reply:
afaik, this way to wrap the model, the root module not being a FSDP model but having model parallelism -- is more close to the real use case.